clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -1994,8 +1995,10 @@ TemplateDeclInstantiator::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
   // Link the instantiation back to the pattern *unless* this is a
   // non-definition friend declaration.
   if (!InstTemplate->getInstantiatedFromMemberTemplate() &&
-      !(isFriend && !D->getTemplatedDecl()->isThisDeclarationADefinition()))
+      !(isFriend && !D->getTemplatedDecl()->isThisDeclarationADefinition())) {
+    InstTemplate->setInstantiatedFromDefinition();
```

Comment:
That's true, but the only problematic cases are the friends.

We could make the flag more accurate, as named, for non-friends, but then we don't have a motivation, and we would make searching for a declaration which is compatible with the definition slightly more expensive for those.

Another option is to rename the flag to something like `InstantiatedFromDefinitionForFriend`.

Reply:
> That's true, but the only problematic cases are the friends.

I was thinking if we can move it to a separate if branch then, for example it might not be conditioned on `isFriend`

```cpp
if (D->getTemplatedDecl()->isThisDeclarationADefinition())
  InstTemplate->setInstantiatedFromDefinition();
```

(I presume a new if won't impact anything other than friends, as you said the friends are the only problematic ones.)